File: linear_multistep/driver.cpp
/* Main routine to test a set of linear multistep methods on the
   scalar-valued ODE problem
     y' = lambda*y + (1-lambda)*cos(t) - (1+lambda)*sin(t), t in [0,5],
     y(0) = 1.

   D.R. Reynolds
   Math 6321 @ SMU
   Fall 2020  */

#include <iostream>
#include "lmm.hpp"

using namespace std;
using namespace arma;


// Define classes to compute the ODE RHS function and its Jacobian

//    ODE RHS function class -- instantiates a RHSFunction
class MyRHS: public RHSFunction {
public:
  double lambda;                              // stores some local data
  int Evaluate(double t, vec& y, vec& f) {    // evaluates the RHS function, f(t,y)
    f(0) = lambda*y(0) + (1.0-lambda)*cos(t) - (1.0+lambda)*sin(t);
    return 0;
  }
};

//    ODE RHS Jacobian function class -- instantiates a RHSJacobian
class MyJac: public RHSJacobian {
public:
  double lambda;                              // stores some local data
  int Evaluate(double t, vec& y, mat& J) {    // evaluates the RHS Jacobian, J(t,y)
    J(0,0) = lambda;
    return 0;
  }
};


// Convenience function for analytical solution
vec ytrue(const double t) {
  vec yt(1);
  yt(0) = sin(t) + cos(t);
  return yt;
};


// main routine
int main() {

  // time steps to try
  vec h("0.1, 0.05, 0.01, 0.005, 0.001");

  // storage for errors
  vec e(h.n_elem);

  // lambda values to try
  vec lambdas("-1.0, -10.0, -50.0, -1000.0");

  // set problem information
  vec y0("1.0");
  double t0 = 0.0;
  double Tf = 5.0;

  // set desired output times
  int Nout = 6;  // includes initial condition
  vec tspan = linspace(t0, Tf, Nout);

  // create ODE RHS and Jacobian objects
  MyRHS rhs;
  MyJac Jac;

  // create true solution results
  mat Ytrue(1,Nout);
  for (size_t i=0; i<Nout; i++) {
    Ytrue.col(i) = ytrue(tspan(i));
  }

  ////////// create linear multistep methods //////////
  string method_names[9];
  int initial_conditions[9];
  bool implicit[9];
  int imethod=0;

  // Adams-Bashforth-2
  method_names[imethod] = "AB-2";
  vec AB2_a("1.0, -1.0, 0.0");
  vec AB2_b("0.0, 1.5, -0.5");
  initial_conditions[imethod] = 2;
  implicit[imethod++] = false;
  LMMStepper AB2(rhs, Jac, y0, AB2_a, AB2_b);

  // Adams-Bashforth-3
  method_names[imethod] = "AB-3";
  vec AB3_a("1.0, -1.0, 0.0, 0.0");
  vec AB3_b(4);
  AB3_b(0) = 0.0;
  AB3_b(1) = 23.0/12.0;
  AB3_b(2) = -16.0/12.0;
  AB3_b(3) = 5.0/12.0;
  initial_conditions[imethod] = 3;
  implicit[imethod++] = false;
  LMMStepper AB3(rhs, Jac, y0, AB3_a, AB3_b);

  // Adams-Bashforth-4
  method_names[imethod] = "AB-4";
  vec AB4_a("1.0, -1.0, 0.0, 0.0, 0.0");
  vec AB4_b(5);
  AB4_b(0) = 0.0;
  AB4_b(1) = 55.0/24.0;
  AB4_b(2) = -59.0/24.0;
  AB4_b(3) = 37.0/24.0;
  AB4_b(4) = -9.0/24.0;
  initial_conditions[imethod] = 4;
  implicit[imethod++] = false;
  LMMStepper AB4(rhs, Jac, y0, AB4_a, AB4_b);

  // Adams-Moulton-2
  method_names[imethod] = "AM-2";
  vec AM2_a("1.0, -1.0");
  vec AM2_b("0.5, 0.5");
  initial_conditions[imethod] = 1;
  implicit[imethod++] = true;
  LMMStepper AM2(rhs, Jac, y0, AM2_a, AM2_b);

  // Adams-Moulton-3
  method_names[imethod] = "AM-3";
  vec AM3_a("1.0, -1.0, 0.0");
  vec AM3_b(3);
  AM3_b(0) = 5.0/12.0;
  AM3_b(1) = 8.0/12.0;
  AM3_b(2) = -1.0/12.0;
  initial_conditions[imethod] = 2;
  implicit[imethod++] = true;
  LMMStepper AM3(rhs, Jac, y0, AM3_a, AM3_b);

  // Adams-Moulton-4
  method_names[imethod] = "AM-4";
  vec AM4_a("1.0, -1.0, 0.0, 0.0");
  vec AM4_b(4);
  AM4_b(0) = 9.0/24.0;
  AM4_b(1) = 19.0/24.0;
  AM4_b(2) = -5.0/24.0;
  AM4_b(3) = 1.0/24.0;
  initial_conditions[imethod] = 3;
  implicit[imethod++] = true;
  LMMStepper AM4(rhs, Jac, y0, AM4_a, AM4_b);

  // BDF-2
  method_names[imethod] = "BDF-2";
  vec BDF2_a(3), BDF2_b(3);
  BDF2_a(0) = 1.0;
  BDF2_a(1) = -4.0/3.0;
  BDF2_a(2) = 1.0/3.0;
  BDF2_b.fill(0.0);
  BDF2_b(0) = 2.0/3.0;
  initial_conditions[imethod] = 2;
  implicit[imethod++] = true;
  LMMStepper BDF2(rhs, Jac, y0, BDF2_a, BDF2_b);

  // BDF-3
  method_names[imethod] = "BDF-3";
  vec BDF3_a(4), BDF3_b(4);
  BDF3_a(0) = 1.0;
  BDF3_a(1) = -18.0/11.0;
  BDF3_a(2) = 9.0/11.0;
  BDF3_a(3) = -2.0/11.0;
  BDF3_b.fill(0.0);
  BDF3_b(0) = 6.0/11.0;
  initial_conditions[imethod] = 3;
  implicit[imethod++] = true;
  LMMStepper BDF3(rhs, Jac, y0, BDF3_a, BDF3_b);

  // BDF-4
  method_names[imethod] = "BDF-4";
  vec BDF4_a(5), BDF4_b(5);
  BDF4_a(0) = 1.0;
  BDF4_a(1) = -48.0/25.0;
  BDF4_a(2) = 36.0/25.0;
  BDF4_a(3) = -16.0/25.0;
  BDF4_a(4) = 3.0/25.0;
  BDF4_b.fill(0.0);
  BDF4_b(0) = 12.0/25.0;
  initial_conditions[imethod] = 4;
  implicit[imethod++] = true;
  LMMStepper BDF4(rhs, Jac, y0, BDF4_a, BDF4_b);

  // combine method objects into an array
  LMMStepper methods[9] = {AB2, AB3, AB4, AM2, AM3, AM4, BDF2, BDF3, BDF4};


  // loop over methods
  for (imethod = 0; imethod<9; imethod++) {

    // set shortcut variables for this method
    string mname = method_names[imethod];
    int k = initial_conditions[imethod];
    bool impl = implicit[imethod];

    // output method name
    cout << endl << mname << ":\n";

    // update solver parameters for implicit methods
    if (implicit[imethod]) {
      methods[imethod].newt.tol = 1.e-3;
      methods[imethod].newt.maxit = 20;
      methods[imethod].newt.show_iterates = false;
    }

    // loop over lambda values
    for (int il=0; il<lambdas.n_elem; il++) {

      // set current lambda value into rhs and Jac objects
      cout << "  lambda = " << lambdas(il) << ":\n";
      rhs.lambda = lambdas(il);
      Jac.lambda = lambdas(il);

      // loop over time step sizes
      for (int ih=0; ih<h.n_elem; ih++) {

        // set initial conditions (backward in time)
        mat y0_LMM(1,k);
        for (int iic=0; iic<k; iic++)
          y0_LMM.col(iic) = ytrue(t0-iic*h(ih));

        // call stepper
        mat Y = methods[imethod].Evolve(tspan, h(ih), y0_LMM);

        // output solution, errors, and overall error
        mat Yerr = abs(Y-Ytrue);
        e(ih) = Yerr.max();
        if (implicit[imethod]) {
          cout << "    h = " << setw(6) << h(ih)
               << "   steps = " << setw(5) << methods[imethod].nsteps
               << "   NIters = " << setw(5) << methods[imethod].nnewt
               << "   max err = " << setw(10) << e(ih);
        } else {
          cout << "    h = " << setw(6) << h(ih)
               << "   steps = " << setw(5) << methods[imethod].nsteps
               << "   max err = " << setw(10) << e(ih);
        }
        if (ih > 0) {
          cout << "   conv rate = " << log(e(ih)/e(ih-1))/log(h(ih)/h(ih-1)) << endl;
        } else {
          cout << endl;
        }

      }
    }
  }

  return 0;
}

File: linear_multistep/driver.py
#!/usr/bin/env python3
#
# Main routine to test various DIRK and IRK methods on the
# scalar-valued ODE problem
#    y' = lambda*y + (1-lambda)*cos(t) - (1+lambda)*sin(t), t in [0,5],
#    y(0) = 1.
#
# D.R. Reynolds
# Math 6321 @ SMU
# Fall 2023
import numpy as np
import sys
sys.path.append('..')
from shared.ImplicitSolver import *
from Explicit_LMM import *
from Implicit_LMM import *

# problem time interval and parameters
t0 = 0.0
tf = 5.0

# flag to switch between dense and iterative linear solvers
iterative = False

# problem-defining functions
def ytrue(t):
    """ Generates a numpy array containing the true solution to the IVP at a given input t. """
    return np.array([np.sin(t) + np.cos(t)])
def f(t,y,lam):
    """ Right-hand side function, f(t,y), for the IVP """
    return np.array([lam*y[0] + (1.0-lam)*np.cos(t) - (1.0+lam)*np.sin(t)])
def J(t,y,lam):
    """ Jacobian (in dense matrix format) of the right-hand side function, J(t,y) = df/dy """
    return np.array( [ [lam] ] )
def Jv(t,y,v,lam):
    """ Jacobian (in dense matrix format) of the right-hand side function, J(t,y) = df/dy """
    return np.array( [lam*v[0]] )

# construct implicit solver
if (iterative):
    solver = ImplicitSolver(Jv, solver_type='gmres', maxiter=20, rtol=1e-9, atol=1e-12)
else:
    solver = ImplicitSolver(J, solver_type='dense', maxiter=20, rtol=1e-9, atol=1e-12, Jfreq=2)

# shared testing data
Nout = 6   # includes initial condition
tspan = np.linspace(t0, tf, Nout)
Ytrue = np.zeros((Nout, 1))
for i in range(Nout):
    Ytrue[i,:] = ytrue(tspan[i])
lambdas = np.array( (-1.0, -10.0, -50.0, -100.0) )
hvals = np.array( (0.1, 0.05, 0.01, 0.005, 0.001) )
errs = np.zeros(hvals.size)

# test runner function
def RunTest(stepper, prevsteps, name, implicit):

    print("\n", name, " tests:", sep='')
    # loop over stiffness values
    for lam in lambdas:

        print("  lambda = " , lam, ":", sep='')
        for idx, h in enumerate(hvals):
            print("    h = %.3f:" % (h), sep='', end='')
            stepper.reset()
            if (implicit):
                stepper.sol.reset()
            # create overly-long initial condition vector (sufficient for all methods)
            y0 = np.zeros((prevsteps+1, ytrue(t0).size), dtype=float)
            y0[-1,:] = ytrue(t0)
            for k in range(1,prevsteps+1):
                y0[-1-k,:] = ytrue(t0-k*h)
            # Note that this is where we provide the rhs function parameter lam -- the "," is
            # required to ensure that args is an iterable (and not a float).
            Y, success = stepper.Evolve(tspan, y0, h, args=(lam,))
            Yerr = np.abs(Y-Ytrue)
            errs[idx] = np.linalg.norm(Yerr,np.inf)
            if (success):
                if (implicit):
                    print("  steps = %4i  Niters = %6i  NJevals = %5i  abserr = %8.2e" %
                          (stepper.get_num_steps(), stepper.sol.get_total_iters(),
                           stepper.sol.get_total_setups(), errs[idx]))
                else:
                    print("  steps = %4i  Nrhs = %6i  abserr = %8.2e" %
                          (stepper.get_num_steps(), stepper.get_num_rhs(), errs[idx]))
        orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
        print('    estimated order:  max = %.2f,  avg = %.2f' %
              (np.max(orders), np.average(orders)))


# Adams-Bashforth-1
alphas, betas, p = AdamsBashforth1()
AB1 = Explicit_LMM(f, alphas, betas)
RunTest(AB1, 0, 'Adams-Bashforth-1', False)

# Adams-Bashforth-2
alphas, betas, p = AdamsBashforth2()
AB2 = Explicit_LMM(f, alphas, betas)
RunTest(AB2, 1, 'Adams-Bashforth-2', False)

# Adams-Bashforth-3
alphas, betas, p = AdamsBashforth3()
AB3 = Explicit_LMM(f, alphas, betas)
RunTest(AB3, 2, 'Adams-Bashforth-3', False)

# Adams-Bashforth-4
alphas, betas, p = AdamsBashforth4()
AB4 = Explicit_LMM(f, alphas, betas)
RunTest(AB4, 3, 'Adams-Bashforth-4', False)

# Adams-Moulton-1
alphas, betas, p = AdamsMoulton1()
AM1 = Implicit_LMM(f, solver, alphas, betas)
RunTest(AM1, 0, 'Adams-Moulton-1', True)

# Adams-Moulton-2
alphas, betas, p = AdamsMoulton2()
AM2 = Implicit_LMM(f, solver, alphas, betas)
RunTest(AM2, 0, 'Adams-Moulton-2', True)

# Adams-Moulton-3
alphas, betas, p = AdamsMoulton3()
AM3 = Implicit_LMM(f, solver, alphas, betas)
RunTest(AM3, 1, 'Adams-Moulton-3', True)

# Adams-Moulton-4
alphas, betas, p = AdamsMoulton4()
AM4 = Implicit_LMM(f, solver, alphas, betas)
RunTest(AM4, 2, 'Adams-Moulton-4', True)

# BDF-1
alphas, betas, p = BDF1()
BDF_1 = Implicit_LMM(f, solver, alphas, betas)
RunTest(BDF_1, 0, 'BDF-1', True)

# BDF-2
alphas, betas, p = BDF2()
BDF_2 = Implicit_LMM(f, solver, alphas, betas)
RunTest(BDF_2, 1, 'BDF-2', True)

# BDF-3
alphas, betas, p = BDF3()
BDF_3 = Implicit_LMM(f, solver, alphas, betas)
RunTest(BDF_3, 2, 'BDF-3', True)

# BDF-4
alphas, betas, p = BDF4()
BDF_4 = Implicit_LMM(f, solver, alphas, betas)
RunTest(BDF_4, 3, 'BDF-4', True)
